### PR TITLE
Allow TextWrapping in combination with TextTrimming

### LINKED ISSUE #4256: TextWrapping doesn't work when using TextTrimming
When using `TextWrapping` and `TextTrimming` at the same time, the text never wraps. On WPF it works as expected.

## Repro

```xaml
<TextBlock MaxWidth="40" TextWrapping="Wrap" TextTrimming="CharacterEllipsis" Text="Hello World! Hello World!" />
```

### Avalonia
avalonia

### WPF
wpf

### PR DESCRIPTION
**Pull request by Gillibald**
## What does the pull request do?
This PR allows a combination of TextWrapping and TextTrimming during text formatting. Now a line is wrapped first and if the line has overflowed it is trimmed.

## What is the current behavior?
Only text trimming or text wrapping is possible not both at the same time.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Fixed issues
Fixes: #4256
#4269

### CODE REVIEW

File: src/Avalonia.Controls/TextBlock.cs [hunk 1]
```diff
@@ -106,7 +106,7 @@ public class TextBlock : Control
         /// Defines the <see cref="TextWrapping"/> property.
         /// </summary>
         public static readonly StyledProperty<TextWrapping> TextWrappingProperty =
-            AvaloniaProperty.Register<TextBlock, TextWrapping>(nameof(TextWrapping));
+            AvaloniaProperty.Register<TextBlock, TextWrapping>(nameof(TextWrapping), TextWrapping.NoWrap);
```
Comment (Gillibald) [inline src/Avalonia.Controls/TextBlock.cs:109]: Revert this. NoWrap is the default

File: samples/ControlCatalog/Pages/TextBlockPage.xaml [hunk 2]
```diff
@@ -18,8 +18,8 @@
       </StackPanel.Styles>
       <Border>
         <StackPanel Width="200" Spacing="8">
-          <TextBlock TextTrimming="CharacterEllipsis" Text="Lorem ipsum dolor sit amet, consectetur adipiscing elit."/>
-          <TextBlock TextTrimming="WordEllipsis" Text="Lorem ipsum dolor sit amet, consectetur adipiscing elit."/>
+          <TextBlock TextWrapping="NoWrap" TextTrimming="CharacterEllipsis" Text="Lorem ipsum dolor sit amet, consectetur adipiscing elit."/>
```
Comment (Gillibald) [inline samples/ControlCatalog/Pages/TextBlockPage.xaml:21]: Revert this

**Comment by jp2masa** [comment]
I tested this and when I set `TextWrapping` to `Wrap` or `WrapWithOverflow` it behaves the same. For example it breaks "abc-def" on the dash, but if it's a long string without break opportunities it doesn't.

**Comment by jp2masa** [comment]
Also, this repro may be useful (it also shows a problem with `MaxHeight` which should be fixed in the future I believe):

```xaml
<TextBlock MaxHeight="40" MaxWidth="40" TextWrapping="Wrap" TextTrimming="CharacterEllipsis" Text="Looooooooooooooong String" />
<TextBlock MaxHeight="40" MaxWidth="40" TextWrapping="Wrap" TextTrimming="CharacterEllipsis" Text="abc-def-ghi-jkl" />
<TextBlock MaxHeight="40" MaxWidth="40" TextWrapping="WrapWithOverflow" TextTrimming="CharacterEllipsis" Text="Looooooooooooooong String" />
<TextBlock MaxHeight="40" MaxWidth="40" TextWrapping="WrapWithOverflow" TextTrimming="CharacterEllipsis" Text="abc-def-ghi-jkl" />
```

### WPF
wpf

### Avalonia
avalonia

**Comment by Gillibald** [comment]
The TextLayout includes the TextLines that overflowed. That makes sense for the case when the first line is affected but it shouldn't be done for more than one line.

**Comment by Gillibald** [comment]
WPF's Trimming behavior for long strings can't be implemented with the current way the TextFormatter works. This needs additional API that moves the collapsing logic into the TextLines implementation and let's the consumer collapse on demand.

**Comment by Gillibald** [comment]
Okay. I will try my best to get all the required API to mimic WPF's behavior into the release. This is a combination of fixing a bug and adding a feature so it should be fine.

**Comment by Gillibald** [comment]
Latest changes should now follow WPF's behavior. Available via build `9255` Please let me know if you find any issues.

**Comment by jp2masa** [comment]
This looks really good, no major issues, thanks for all your work!

The only issue I could find is not related to WPF, but `TextTrimming` doesn't work with `MaxLines`, so for example replacing `MaxHeight="40"` with `MaxLines="2"` will not show the ellipsis character. I hope this one isn't hard to fix.

**Comment by Gillibald** [comment]
I will apply the same logic on MaxLines tomorrow

**Comment by Gillibald** [comment]
Should now handle MaxLines properly. Build for testing `9271`

**Comment by jp2masa** [comment]
Thanks! It works perfectly!

**Review by danwalmsley — APPROVED** [review]
LGTM

**Pull request merged**